### ISSUE TITLE
Package ZPT-Sharp for Debian

### ISSUE THREAD

**Issue by craigfowler** [csf-dev/ZPT-Sharp#85]
This Debian package should be automatically generated by the 'Deploy' build process. It must conform with the DFSG and be suitable for upload to the 'Main' Debian repository.

## Blocked by
This is currently blocked by all of the following tickets:

* #95 
* #187 
* #188
* #189
* #190
* #191
* #192
* #193 

## Planned packages
I would like to split ZPT-Sharp into a number of Debian packages, in a very similar manner to how the Nuget packages are organised:

### Packages with actual content
These packages have actual files included.  Each one installs only one or two DLL files.

* **libzptsharp-CSF.Zpt** - *depends on abstractions package*
* **libzptsharp-CSF.Zpt.Abstractions**
* **libzptsharp-CSF.Zpt.Cli** - *depends on core API package*
* **libzptsharp-CSF.Zpt.MVC** - *depends on core API package*
* **libzptsharp-CSF.Zpt.DocumentProviders.HtmlHAP** - *depends on core API package*
* **libzptsharp-CSF.Zpt.DocumentProviders.XmlLinq** - *depends on core API package*
* **libzptsharp-CSF.Zpt.DocumentProviders.XmlLegacy** - *depends on core API package*
* **libzptsharp-CSF.Zpt.ExpressionEvaluators.CSharpExpressions** - *depends on core API package*
* **libzptsharp-CSF.Zpt.ExpressionEvaluators.NotExpressions** - *depends on core API package*
* **libzptsharp-CSF.Zpt.ExpressionEvaluators.PathExpressions** - *depends on core API package*
* **libzptsharp-CSF.Zpt.ExpressionEvaluators.PythonExpressions** - *depends on core API package*
* **libzptsharp-CSF.Zpt.ExpressionEvaluators.StringExpressions** - *depends on core API package*

Each of these packages also needs two counterpart packages:

* A `-dbg` package which contains the debugging symbols
* A `-doc` package which contains the XML documentation

Also, there needs to be a **libzptsharp-src** package which includes all of the source code (for all of ZPT-Sharp).

### Meta-packages
*AKA virtual packages in the Debian world*.  These packages only have dependencies on other packages, in order to group them together for a single installation.

* **libzptsharp-base** - *depends on the core API package and each of the "default plugins" just as with the Nuget packages*
* **zptsharp-builder** - *depends on the CLI package and `libzptsharp-base` in order to provide the ZptBuilder app*
* **zptsharp-mvc5** - *depends on the MVC package and `libzptsharp-base` in order to provide a working MVC view engine for ASP.NET web apps*

**Comment by craigfowler** [comment]
https://www.debian.org/doc/manuals/maint-guide/first.en.html

**Comment by craigfowler** [comment]
A big consideration is the Debian Free Software Guidelines.

The HTML agility pack is licenced under the MS-PL. This licence is not dfsg-compliant. This means it would need to go into *contrib* and not *main*.

What I would have to do is construct a series of Debian packages:
* libzptsharp-core - This is the abstractions and main libraries. Packaged in *main* as they have no dependencies of its own.
* libzptsharp-xml - This is the two XML implementations for ZPT pages. Can go in *main* because they have no external dependencies. Depends upon core though.
* zptsharp-builder - This is the builder app. It depends upon core but can still go into *main*.
* libzptsharp-htmlhap - This is the HTML HAP-based implementation. Has to go into *contrib* because it has to be MS-PL licenced.
* libzptsharp-mvc5 - This is the MVC 5 view engine. Need to investigate the licencing of its dependencies (the nuget packages for MVC 5). If they are MS-PL then this also needs to be, thus it goes into *contrib*. Otherwise it might be able to go into *main*.

**Comment by craigfowler** [comment]
Actually, that's not **contrib**, it's **non-free** where those packages go. Contrib is for things that are dfsg-compliant but have dependencies on things which aren't (for build or execution).

**Comment by craigfowler** [comment]
Perhaps the above is not true. Maybe the MS-PL is indeed dfsg-compliant? Precedent is that parts of mono itself is distributed in Debian, and some of its source is MS-PL. If it's in main then I see no reason why ZPT-sharp can't go into main?

**Comment by craigfowler** [comment]
Here's initial work on a script to create a Debian package.  It's nowhere near complete and probably needs throwing out.  I'm going to remove it from source control since it's likely not worth keeping (in its present form).  But - here it is for sake of archival:

```bash
#!/bin/bash

SCRIPT_DIR="$(dirname "$0")"
ROOT_DIR="${SCRIPT_DIR}/.."
BUILD_DIR="DEBIAN"

PACKAGE_NAME="zptsharp-zptbuilder"

cd "${ROOT_DIR}/CSF"
git archive HEAD -o ../CSF.tar.gz
cd ..

git archive HEAD -o CSF.Zpt.tar.gz

rm -rf "$BUILD_DIR"
mkdir -p "$BUILD_DIR"
cd "$BUILD_DIR"

tar -xvf ../CSF.Zpt.tar.gz
cd CSF
tar -xvf ../../CSF.tar.gz
cd ..

PACKAGE_VERSION="$(egrep "AssemblyVersion" Common/CommonAssemblyInfo.cs | egrep -o "[0-9.]+")"

nuget restore
xbuild
```

**Comment by craigfowler** [comment]
Closing as WONTFIX. Packaging ZPT-Sharp for Debian is currently not a priority. This can be reopened if it ever becomes a priority.

NuGet is good enough at the moment.
